File: 7-two-sum/two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        
        // Naive implementation
        // Time complexity: O(n^2)
        
        /*
        for (int i = 0; i < nums.length; i++) 
            for (int j = 0; j < nums.length; j++) 
                if (nums[i] + nums[j] == target && i != j) {
                    int[] ans = {i, j};
                    return ans;
                }
        return null;
        */
        
        // Hashmap implementation
        // Time complexity: O(n)
        
        Map<Integer, Integer> table = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int diff = target - nums[i];
            if (table.containsKey(diff)) {
                return new int[] {table.get(diff), i};
            }
            table.put(nums[i], i);
        }
        return null;
    }
}

File: 7-two-sum/two-sum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        
        # Naive implementation: using nested for loop
        # Time complexity: O(n^2)
        
        """
        for i in range(len(nums)):
            for j in range(len(nums)):
                if nums[i] + nums[j] == target and i != j:
                    return (i, j)
        """
                
        # Hashtable implementation
        # Time complexity: O(n)
        
        table = {}
        for count, value in enumerate(nums):
            diff = target - value
            if diff in table:
                return [table[diff], count]
            table.update({value:count})
